Make CTE Materialization the Default Instead of Inlining

kryonix opened the pull request:
Until now, DuckDB's default behavior for CTEs was to inline them, which means that
the CTE is replaced with its definition in the main query. However, inlining can not
only lead to performance issues, especially when the CTE is used multiple times in
the main query, but also to incorrect results in some cases.

With the introduction of the `WITH ... AS [NOT] MATERIALIZED` syntax in https://github.com/duckdb/duckdb/pull/7126,
users can specify that a CTE should be materialized, meaning that it will be 
computed once and stored in memory for the duration of the main query.
This can improve performance and ensure correct results when the CTE is used multiple times.

This PR is a continuation of that work. While we still want to be able to inline CTEs in some cases,
it should not be the default behavior anymore. More importantly, this should not be the responsibility
of the `Binder`, but rather an `Optimizer` decision.

For CTEs with `MATERIALIZED` or `NOT MATERIALIZED` specified, nothing changes.
If a CTE is marked as `MATERIALIZED`, it will be materialized, and if it is marked
as `NOT MATERIALIZED`, it will be inlined—regardless of the possibly questionable
performance or correctness implications.

For CTEs without a `MATERIALIZED` or `NOT MATERIALIZED` clause, the `Optimizer` will
decide whether to inline or materialize the CTE.

The rules for this decision are currently a work in progress.

---

**Performance**

@neumannt released a solution of the Advent of Code 2024 using SQL: https://github.com/neumannt/aoc24
All solutions rely heavily on CTEs, but do not use explicit MATERIALIZATION hints.
I use his queries as a benchmark—using the full AoC input:

| DAY | Speedup with this PR vs. DuckDB main |
|-----|---------------------------------------|
| 1   |  2.3x                                 |
| 2   |  1.7x                                 |
| 3   |  1.6x                                 |
| 4   |  2.7x                                 |
| 5   | 99.3x                                 |
| 6   | 1810x (now ~18 sec, main ~9h)               |
| 7   | 11.8x                                 |
| 8   |  1.2x                                 |
| 9   |156.2x                                 |
| 10  |  1.7x                                 |
| 11  |  1.2x                                 |
| 12  |  2.3x                                 |
| 13  |  3.2x                                 |
| 14  |  308x (now ~1.5s, main ~8 min)               |
| 15  | > 10x (now ~2 min, main timeout)       |
| 16  | TBD                                   |
| 17  | Not supported                         |
| 18  | > 73x (now ~8.5 sec, main timeout)     |
| 19  |  2.2x                                 |
| 20  | 28x (now ~8 sec, main ~3 min 50s)           |
| 21  | 24.2x                                 |
| 22  | Not supported                         |
| 23  | 188x (now ~260ms, main ~48 sec)               |
| 24  | Not supported                         |
| 25  | > 2245x (now ~250ms, main timeout)     |

I think these numbers speak for themselves. All queries experienced a speedup, and
the most problematic queries are now fast enough to run in a reasonable time frame.
Speedups range from 1.2x to over 2000x, which—I think—is a significant improvement.

---

`logical_operator_deep_copy` is a new visitor that allows us to _"deep copy"_ a logical operator tree.
While we can already get a 1:1 copy of a logical operator using `Copy`, we can not use that
copy to immediately replace a CTE reference in the main query. This is because the copy
has the same `table_index`es as the original tree, which breaks some sanity checks.
To fix this, we traverse the tree and assign new `table_index`es to each operator.

---

TODO:
- [x] Improve the rules for inlining vs. materializing CTEs
- [x] Fix the unit tests for `WITH RECURSIVE` queries, that are currently failing because they relied on the inlining behavior
- [x] Clean up some rough edges in the code
- [x] Add more tests for the new `Optimizer` behavior

---

Fix: #17456
Fix: https://github.com/duckdb/duckdb/issues/12875
Fix: https://github.com/duckdb/duckdb/issues/15416
Fix: https://github.com/duckdb/duckdb/issues/16891
Fix: https://github.com/duckdb/duckdb/issues/9940
Fix: https://github.com/duckdb/duckdb/issues/14244

kryonix committed:
Move CTE inlining from the Binder to the Optimizer

CTE inlining has been done in the Binding-Phase since their
introduction. That, however, is very early and can lead to issues.

Add logical_operator_deep_copy module that creates a deep copy of
logical plans. The plan is copied as usual first. Then all table indexes
are renewed using the Binder. Finally, all table access expressions are
fixed to use the new bindings.

kryonix committed:
Set materialized field of BoundCTENode based on CTENode

kryonix committed:
Add test for issue 15416

kryonix committed:
Make format-fix

kryonix committed:
Remove some bogous comments in cte_inlining and dead code in bind_basetableref

szarnyasg commented:
Added #17456 to the list this is (hopefully) fixing

soerenwolfers commented:
> if it is marked
as `NOT MATERIALIZED`, it will be inlined—regardless of the possibly questionable
performance or correctness implications

Is that because it's difficult to get the "correct" behavior and this will be tackled later, or because it has been decided that the remaining deviations from PostgreSQL (where materialization is a pure performance question and cannot override semantic necessities) in the presence of `NOT MATERIALIZED` are desired (concretely, lateral column reference resolution and repeated execution of volatile functions)? 

Asking because this is a very frequent source of issues / confusion and I don't want to give people false advice / info on the status and future of this.

kryonix commented:
> Is that because it's difficult to get the "correct" behavior and this will be tackled later, or because it has been decided that the remaining deviations from PostgreSQL (where materialization is a pure performance question and cannot override semantic necessities) in the presence of `NOT MATERIALIZED` are desired (concretely, lateral column reference resolution and repeated execution of volatile functions)?
> 
> Asking because this is a very frequent source of issues / confusion and I don't want to give people false advice / info on the status and future of this.

Great question! My thinking is as follows, consider the following query:

```
WITH questionable(x) AS NOT MATERIALIZED (
    SELECT random()
) SELECT * FROM questionable q1, questionable q2;
```

Clearly, inlining duplicates the `random()` expression—leading to, well, questionable semantics. To get the "correct" behavior, we would have to ignore the `NOT MATERIALIZED` annotation, and materialize the CTE regardless. This is where I tend to have a different opinion than the PostgreSQL devs. But I appreciate a discussion here. I think that DuckDB should not overrule the user's decision to inline/not inline the CTE. My argument is, that `[NOT] MATERIALIZED` essentially is a _plan hint_, "trust me, I know what I'm doing".

I think the question boils down to: Should the system allow `[NOT] MATERIALIZED` annotations, regardless of seemingly wrong semantics (after inlining)?

--- 

My current goal with this PR is more to fix the behavior of CTEs without any `[NOT] MATERIALIZED` annotations.

kryonix commented:
> Something I recently discovered that I thought was a nice feature was the support for running mutually recursive queries, something that, even though is not supported per se, is made possible due to CTE inlining as described in #14716
> 
> I wanted to chime in here to point out that in these cases it might not be an optimization decision to inline those CTEs, and so I agree with:
> 
> > My argument is, that [NOT] MATERIALIZED essentially is a plan hint, "trust me, I know what I'm doing".
> 
> I wonder if by changing the responsibility of inlining from the parser to the optimizer, this "feature" will no longer work since the parser might throw an error with these types of queries?

The way this PR changes the behavior does not affect this, let's call it "happy little accident" 😉—at least currently.

However, I do agree with @soerenwolfers, that the "variable leaking" is a huge problem, and should be prohibited. This plain and simple is a bug and needs fixing—there are several issues regarding this (hope I've listed most of them above).
> should still not allow "variable leaking" (like PG)

Problem is: The easiest way of doing that would be to remove the inlining code path entirely from the Binder, which would then also prohibit "mutually recursive queries".

mikegarciam commented:
> However, I do agree with @soerenwolfers, that the "variable leaking" is a huge problem, and should be prohibited. This plain and simple is a bug and needs fixing—there are several issues regarding this (hope I've listed most of them above).

Agreed 100%. 
I wonder if this leaking behavior has more to do with implicit `LATERAL` joins rather than CTE inlining itself. From what I could see in the [docs](https://duckdb.org/docs/stable/sql/query_syntax/from.html):
```
DuckDB detects when LATERAL joins should be used, making the use of the LATERAL keyword optional.
```

and there is currently no way of preventing the automatic lateral join. 

Would forcing explicit lateral joins (which I believe is more aligned with PG syntax) prevent this leakage?

kryonix commented:
> Would forcing explicit lateral joins (which I believe is more aligned with PG syntax) prevent this leakage?

Technically yes, I believe. However, I think this would just shift the problem to another place. Because then, the following query would still be bogous:

```SQL
WITH t(x) AS (SELECT x)
SELECT *
FROM  range(10) AS _(x), LATERAL (SELECT * FROM t);
```

Maybe we could address this then, sure. But it does not really feel like a definitive solution. Now, having said that. Beyond changing the default behavior of CTEs (this PR) I did not really look into different ways to fix this standing issue. I think a better way of keeping "mutually recursive CTEs" would be to _do it right_, and add it as a feature. This is beyond the scope of this PR, of course. But then, we could have a sane definition and an explicit implementation. Regarding semantics, we would have to discuss (I'm sure @timfi—who tries to convince me of doing just that—has some ideas 😉).

soerenwolfers commented:
Agreed: I think the CTE itself should get a `LATERAL` modifier (defaulting to false).

 The question whether joins should be lateral by default is independent of that. If joins were made non-lateral by default, then that last query from @kyronix would need the `LATERAL` keyword both on the CTE and on the join, otherwise it would need it only on the CTE.

kryonix committed:
Remove some left overs from cte_inlining optimizer

kryonix committed:
Fix name shadowing issue with materialized CTEs

kryonix committed:
Disable verification for some infinite recursive CTEs

kryonix committed:
Adapt test_null_type_propagation with CTEs

kryonix committed:
Add missing import to logical operator deep copy

kryonix committed:
Try CTE inlining for multiple references if scans are below LIMIT

kryonix committed:
Add explicit default to serialization of CTEMaterialize

kryonix committed:
Remove MARK join correlation check---DecorrelateIndependent handles this

kryonix committed:
Fix faulty include in logical deep copy

kryonix committed:
Fix volatile check in cte inlining optimizer

kryonix committed:
Make format-fix

kryonix committed:
Add missing override in deep copy

kryonix committed:
Prevent inlining if CTE definition ends in aggregate or distings

This mimics the behavior of the current automatic materialization in the
binder

kryonix committed:
Fix issue in OrderPreservationRecursive

The OrderPreservationType of materialized CTEs was determined wrong,
because the child_idx was never incremented. This could lead to some
strange behavior.

kryonix committed:
Fix ExtractJoinRelations for materialized CTEs

kryonix committed:
Optimize decorrelation of materialized CTEs

We do not need to push the DependentJoin into the left side
(materialization side) of the CTE, if only the right side is correlated.

kryonix committed:
Push DEPENDENT JOIN below materialized CTE if lhs is free of correlation

We can push the DEPENDENT JOIN below a materialized CTE IFF only the rhs
is correlated.

kryonix committed:
Remove unnecessary move

kryonix committed:
Fix test 1248 recursive CTEs do not necessarily preserve order

kryonix committed:
Improve statistics of CTE and CTE_REF operators

kryonix committed:
Disable now broken test case

Usually that is not the right way of dealing with a test suite. However,
this test can not be unified with the changes made to CTEs.
no_alternative_verify no longer protects us from the non-deterministic
behavior this test case has.

kryonix committed:
Fix use after move in relation_manager.cpp

kryonix committed:
If a CTE requests to never be materialized, we try to inline it

kryonix committed:
Remove CTE inlining code path from the binder

ALL CTE inlining now happens exclusively in the optimizer.

kryonix committed:
Remove dead CTE materialization code from the binder

kryonix committed:
Recreate storage_version.db

kryonix committed:
Fix DetectCorrelatedExpressions for materialized CTEs

kryonix committed:
Simplify DetectCorrelatedExpressions for CTEs a bit

kryonix committed:
Compare CTE column names case insensitive in duplicate check

kryonix committed:
Try to fix QueryRelation::Bind for python replacement scans

kryonix committed:
Fix faulty python TestReplacementScan (test_same_name_cte)

kryonix committed:
Improve logical plan output of CTEs

kryonix committed:
Add a few tests for the cte inlining optimizer

kryonix commented:
I'm quite excited that this PR is now finally ready for review.
This PR does quite a few things—actually more than I initially intended.
Apart from the already known issues listed in the PR description, I have
found a lot more hidden bugs related to CTEs, and I have fixed them all:

- Obviously, it fundametally changes the way we handle CTEs in general.
  *All* inlining of CTEs is now done at a single place, which is the new `CTEInlining` optimization pass.
- The entire CTE inlining logic that previously lived in the `Binder` class has been removed.

  The consequence of that change is that "mutually recursive CTEs" are no longer working.
  I understand that some people might be disappointed by that—and I'm sorry for that—but
  I am convinced that this is the right way to go. Mutually recursive CTEs should be
  implemented explicitly using a different approach, not as a side effect of the CTE
  inlining logic. This was never intended, and even the documentation states that this
  is not supported. I do have some ideas on how to implement mutually recursive CTEs in
  a more explicit way, but I will leave that for a possible future PR.

  However, with this change, free identifiers in CTEs now lead to an error, as they
  should. This has been a long-standing issue and source of confusion, and I am
  very happy that we finally eliminate this problem once and for all.
- We did improve CTE statistics once again, so that we now get a more accurate
  estimate of the number of rows produced by a CTE.
- We have fixed a bug with the `OrderPreservation` determination for CTEs (again..),
  which was not working correctly.
- Decorrelation of materialized CTEs has been improved, and some bugs have been
  fixed. This generally leads to better plans.
- We have removed some faulty tests that were logically incorrect, or have become
  obsolete due to the changes in the CTE handling.
- CTE name shadowing has been improved significantly, so that we now have a much
  more robust handling of CTE names that are shadowed by other CTEs or subqueries.
- The excellent automatic CTE materialization feature by @lnkuiper has been
  integrated into the new CTE inlining pass. The now obsolete code in the `Binder`
  has been removed.
- We have fixed a bug in `bind_cte_node`, that caused duplicate column names to
  not be detected and renamed correctly.
- Materialized CTEs within CTEs were not handled correctly, which has been fixed.
- CTEs with the same name as a dataframe in python were basically ignored, this
  now works correctly and is not an issue anymore.

Surely, there are more changes, but these are the most important ones.

I think the new way we handle CTEs is much more robust, clean, more explicit, and as the
benchmark implies, also potentially leads to (much) better performance in many cases.

kryonix committed:
Fix a few test cases

lnkuiper commented on src/include/duckdb/planner/logical_operator_deep_copy.hpp:
```diff
@@ -0,0 +1,67 @@
+
+private:
+	Binder &binder;
+	std::map<idx_t, idx_t> table_idx_replacements;
```
I don't see the code relying the table indices being ordered. Can this be an `unordered_map` instead?

kryonix replied:
Good point. This is absolutely not required.

lnkuiper commented on src/optimizer/cte_inlining.cpp:
```diff
@@ -0,0 +1,225 @@
+			// this CTE is only referenced once, we can inline it directly without copying
+			bool success = Inline(op->children[1], *op, false);
+			if (success) {
+				op = std::move(op->children[1]);
```
If a CTE contains joins, and the CTE is joined with more tables later in the query, and the CTE is inlined, we could, in theory, perform join ordering on the entire plan. Currently, however, DuckDB will perform join ordering within the CTE, and will reorder the CTE where it is referenced, but does not reorder the the joins within the CTE with the joins outside of the CTE.

With your PR, this is still the case, which is OK, we probably should not increase the scope of this PR. However, if we are somehow able to remove the projection that sits at the top of CTEs, we would be able to perform join ordering on the entire plan. What do you think about this? It would complicate column binding replacement, so maybe it would be a nightmare to implement, but it could improve plans for queries with many CTEs/joins.

Maybe this should be implemented as a separate optimizer pass after this optimizer, which would remove useless projections. Then it would also work for subqueries, not just CTEs. This optimizer could remove projections that only reference columns without actually performing any expressions. We can only do this in places where the projection doesn't affect semantics, so not at the query plan root (and not under set operations maybe?), for example.

kryonix replied:
I absolutely agree! I thought that this would already happen in the optimizer—but I was wrong. The removal of these trivial projections happens during physical plan mapping, which of course is too late for the join order optimizer. I think adding this as an optimization pass is a great idea. And, as you said, would be great for other situations as well. However, I agree that further extending the scope of this PR may be a little bit too much.

Adding the projection during CTE inlining is a nice trick to keep the code simple and the operator indexes unique, while avoiding to change all column references further up the operator tree.

lnkuiper reviewed (commented):
This looks excellent! I just have a few comments:

kryonix committed:
Fix issue #14244

kryonix committed:
Replace map with unordered_map in logical_operator_deep_copy

kryonix committed:
Revert recreating storage_version.db

kryonix commented:
Looks like the CI failure because of the `storage_version.db` is now re-appearing. Based on the error message `Serialization Error: Failed to deserialize: field id mismatch, expected: 102, got: 65535` I suspect that this change causes the error:

https://github.com/duckdb/duckdb/pull/17459/files#diff-b82f2cae7353c89d5df038a37481b1a57d740e23744b7776a9b029b0ba600011R44-R45

I thought adding this field with an explicit default would do the trick, but it seems like that is not the case? Regenerating the `storage_version.db` fixes this, but I know that that's usually not wanted, so I'm not sure if that's the way to go.

lnkuiper commented:
@kryonix Right, that seems like the issue. The original JSON results in this:
```
void CommonTableExpressionInfo::Serialize(Serializer &serializer) const {
	serializer.WritePropertyWithDefault<vector<string>>(100, "aliases", aliases);
	serializer.WritePropertyWithDefault<unique_ptr<SelectStatement>>(101, "query", query);
	serializer.WriteProperty<CTEMaterialize>(102, "materialized", materialized);
	serializer.WritePropertyWithDefault<vector<unique_ptr<ParsedExpression>>>(103, "key_targets", key_targets);
}

kryonix committed:
Revert some changes to the serialization and fix VIEWs

kryonix commented:
You are right, these changes were unnecessary and I just reverted them. This uncovered another bug related to CTEs in VIEWs loaded from the serialized format, which is now also fixed. Let's see if the CI agrees, locally all tests succeeded 😉

lnkuiper reviewed (approved):
LGTM now, thanks!

Mytherin commented:
Awesome, thanks!

alippai commented:
This is cool!
Vertica has similar, there you can specify the inlining limit. By default the CTEs used once are inlined, CTEs used multiple times are materialized.
